### PR TITLE
[TACHYON-481] Tests for the DataServer refactor

### PR DESCRIPTION
**Pull request by gpang**
Several unittests and integration tests for the DataServer.

### CODE REVIEW

File: servers/src/main/java/tachyon/worker/netty/protocol/buffer/DataBuffer.java [hunk 1]
```diff
@@ -33,4 +35,11 @@
    * @return the length of the data in bytes.
    */
   public abstract long getLength();
+
+  /**
+   * Returns a {@link ByteBuffer} for read-only access to the data.
+   *
+   * @return a read-only ByteBuffer representing the data
+   */
+  public abstract ByteBuffer getReadOnlyByteBuffer();
```
Comment (calvinjia) [inline servers/src/main/java/tachyon/worker/netty/protocol/buffer/DataBuffer.java]: Is this intended only for testing use?

File: integration-tests/src/test/java/tachyon/worker/netty/protocol/MessageSavingHandler.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.SimpleChannelInboundHandler;
+
+/**
+ * This is a simple handler for the channel pipelines. This simply saves the message it receives.
+ * The stored message can be retrieved with getMessage() which is blocking until the message is
+ * received.
+ */
+@ChannelHandler.Sharable
+public class MessageSavingHandler extends SimpleChannelInboundHandler<RPCMessage> {
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/worker/netty/protocol/MessageSavingHandler.java]: Is this class mainly for test use? If yes, better to state so in its javadoc

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockRequestTest.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.Unpooled;
+
+public class RPCBlockRequestTest {
+  public static final long BLOCK_ID = 11;
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockRequestTest.java]: Any specific reason these fields are public?

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+public class RPCBlockResponseTest {
+  public static final long BLOCK_ID = 1;
+  public static final long OFFSET = 2;
+  // The RPCMessageEncoder sends the payload separately from the message, so these unit tests only
+  // test the message encoding part. Therefore, the 'length' should be 0.
+  public static final long LENGTH = 0;
+
+  private ByteBuf mBuffer = null;
+
+  private void assertValid(long blockId, long offset, long length, RPCBlockResponse req) {
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: NIT: is resp a better parameter name than req here?

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    Assert.assertEquals(blockId, req.getBlockId());
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: getType() is not unit-tested. Probably you can test it here?

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    assertValid(BLOCK_ID, OFFSET, LENGTH, req);
+    assertValid(BLOCK_ID, OFFSET, LENGTH, req2);
+  }
+
+  @Test
+  public void validateTest() {
+    RPCBlockResponse req = new RPCBlockResponse(BLOCK_ID, OFFSET, LENGTH, null);
+    assertValid(req);
+  }
+}
```
Comment (apc999) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: So far, all the DataBuffer params in constructors are passed as null in this test.
I suggest to also cover testing getPayloadDataBuffer, which takes a non-null DataBuffer in constructor, and test getPayloadDataBuffer() returns the same reference?

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java [hunk 7]
```diff
@@ -56,20 +56,23 @@
   public static Collection<Object[]> data() {
     // creates a new instance of DataServerTest for each network type
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: "for each network type" (and transfer type" ?

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java [hunk 8]
```diff
@@ -107,6 +110,7 @@ private void assertValid(final DataServerMessage msg, final int expectedSize,
   @Before
   public final void before() throws IOException {
     System.setProperty(Constants.WORKER_DATA_SERVER, mDataServerClass);
+    System.setProperty(Constants.WORKER_NETTY_FILE_TRANSFER_TYPE, mNettyTransferType);
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: This line only takes effect if the class is netty server, right?
Can we set the transfer type to be something like NOT USED for NIO server (in line 61)?

File: integration-tests/src/test/java/tachyon/worker/netty/protocol/RPCMessageIntegrationTest.java [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+    sBootstrapClient = new Bootstrap();
+    sBootstrapClient.group(sEventClient);
+    sBootstrapClient.channel(NioSocketChannel.class);
+    sBootstrapClient.handler(new PipelineInitializer(new MessageSavingHandler()));
+  }
+
+  @AfterClass
+  public static void afterClass() {
+    // Shut everything down.
+    sEventClient.shutdownGracefully();
```
Comment (apc999) [inline integration-tests/src/test/java/tachyon/worker/netty/protocol/RPCMessageIntegrationTest.java]: I vaguely remember @calvinjia had PRs before to change shutdownGracefully() to shutdown() before to speed up the tests where we don't need to be graceful :))) Please double check with him if we should do the samething here.

Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/worker/netty/protocol/RPCMessageIntegrationTest.java]: You can pass in arguments to `shutdownGracefully` to make it shutdown faster. I think by default it waits until 2 seconds pass without any new messages, up to 15 seconds if new messages continue to come in.

File: servers/src/main/java/tachyon/worker/netty/protocol/buffer/DataBuffer.java (hunk 1 above)
Comment (gpang) [inline servers/src/main/java/tachyon/worker/netty/protocol/buffer/DataBuffer.java]: Not necessarily. This interface could be used for the client side as well, when that is implemented. I'm not committed to the exactly API, so, it may change slightly when the client is refactored to use netty.

In general, I think there needs to be two ways to read the data: one for Netty output, and one to actually read the bytes.

File: integration-tests/src/test/java/tachyon/worker/netty/protocol/MessageSavingHandler.java (hunk 2 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/worker/netty/protocol/MessageSavingHandler.java]: I added a comment.

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockRequestTest.java (hunk 3 above)
Comment (gpang) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockRequestTest.java]: No, I changed it to private.

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java (hunk 4 above)
Comment (gpang) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: renamed.

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java (hunk 5 above)
Comment (gpang) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: Added getType() assertion.

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java (hunk 6 above)
Comment (gpang) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: added the simple test.

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java (hunk 7 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: clarified the comments.

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java (hunk 8 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: Done.

File: integration-tests/src/test/java/tachyon/worker/netty/protocol/RPCMessageIntegrationTest.java (hunk 9 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/worker/netty/protocol/RPCMessageIntegrationTest.java]: I changed the parameters for faster shutdowns.

**Comment by gpang** [comment]
Thanks @apc999 and @calvinjia for the review! I updated the PR with your feedback.

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java [hunk 10]
```diff
@@ -107,6 +113,9 @@ private void assertValid(final DataServerMessage msg, final int expectedSize,
   @Before
   public final void before() throws IOException {
     System.setProperty(Constants.WORKER_DATA_SERVER, mDataServerClass);
+    if (!mNettyTransferType.equals(UNUSED_TRANSFER_TYPE)) {
+      System.setProperty(Constants.WORKER_NETTY_FILE_TRANSFER_TYPE, mNettyTransferType);
```
Comment (calvinjia) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: I think the System property should be cleared in `after()`? On a more stylistic note, does setting or not setting the netty transfer type matter if its unused (ie. is the if necessary)?

File: servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java (hunk 4 above)
Comment (apc999) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: Maybe I am too picky, in this file, the other variables/parameters of type RPCBlockResponse are still named as req. Better to keep them consistent? :)

Comment (gpang) [inline servers/src/test/java/tachyon/worker/netty/protocol/RPCBlockResponseTest.java]: Sorry, that's my fault. I renamed all the 'req' to 'resp'.

File: integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java (hunk 10 above)
Comment (gpang) [inline integration-tests/src/test/java/tachyon/worker/DataServerIntegrationTest.java]: Thanks for catching that! I added the clear in after(). I also got rid of the 'if' since it looks nicer and shouldn't change any behavior.

**Comment by calvinjia** [comment]
+1 LGTM

**Comment by apc999** [comment]
+1 LGTM
